runtime/src/slots.rs

```diff
@@ -0,0 +1,1425 @@
+						d.remove(0);
+						<Deposits<T>>::insert(id, &d);
+						// Then we need to get the new amount that should continue to be held on
+						// deposit for the parachain.
+						let new_held = d.into_iter().max().unwrap_or_default();
+						// If this is less than what we were holding previously, then return it
+						// to the parachain itself.
+						if let Some(rebate) = outgoing.checked_sub(&new_held) {
+							T::Currency::deposit_creating(
+								&id.into_account(),
+								rebate
+							);
+						}
+						// We keep this entry around until the next lease period.
+						true
+					}
+				} else {
+					false
+				}
+			}).collect::<Vec<_>>();
+			*ids = new;
+		});
+
+		// Deploy any new chains that are due to be commissioned.
+		for para_id in <OnboardQueue<T>>::take(lease_period_index) {
+			if let Some((_, IncomingParachain::Deploy{code, initial_head_data}))
+				= <Onboarding<T>>::get(&para_id)
+			{
+				// The chain's deployment data is set; go ahead and register it, and remove the
+				// now-redundant on-boarding entry.
+				let _ = T::Parachains::register_parachain(para_id.clone(), code, initial_head_data);
+				// ^^ not much we can do if it fails for some reason.
+				<Onboarding<T>>::remove(para_id)
+			}
+		}
+	}
+
+	/// Actually place a bid in the current auction.
+	///
+	/// - `bidder`: The account that will be funding this bid.
+	/// - `auction_index`: The auction index of the bid. For this to succeed, must equal
+	/// the current value of `AuctionCounter`.
+	/// - `first_slot`: The first lease period index of the range to be bid on.
+	/// - `last_slot`: The last lease period index of the range to be bid on (inclusive).
+	/// - `amount`: The total amount to be the bid for deposit over the range.
+	fn handle_bid(
+		bidder: Bidder<T::AccountId, ParaIdOf<T>>,
+		auction_index: u32,
+		first_slot: LeasePeriodOf<T>,
+		last_slot: LeasePeriodOf<T>,
+		amount: BalanceOf<T>
+	) -> Result<(), &'static str> {
+		// Bidding on latest auction.
+		ensure!(auction_index == <AuctionCounter<T>>::get(), "not current auction");
+		// Assume it's actually an auction (this should never fail because of above).
+		let (first_lease_period, _) = <AuctionInfo<T>>::get().ok_or("not an auction")?;
+
+		// Our range.
+		let range = SlotRange::new_bounded(first_lease_period, first_slot, last_slot)?;
+		// Range as an array index.
+		let range_index = range as u8 as usize;
+		// The offset into the auction ending set.
+		let offset = Self::is_ending(<system::Module<T>>::block_number()).unwrap_or_default();
+		// The current winning ranges.
+		let mut current_winning = <Winning<T>>::get(offset)
+			.or_else(|| offset.checked_sub(&One::one()).and_then(<Winning<T>>::get))
+			.unwrap_or_default();
+		// If this bid beat the previous winner of our range.
+		if current_winning[range_index].as_ref().map_or(true, |last| amount > last.1) {
+			// This must overlap with all existing ranges that we're winning on or it's invalid.
```

Comment:
Why does it matter that your bid intersects with another slot you are currently winning?